### PR TITLE
Refactor parser

### PR DESCRIPTION
**Pull request by fmeringdal**
The parser module has been largely rewritten from scratch to be more Rust ergonomic (remember that the module was ported from the JS implementation almost line by line) and use terminology more consistent with the RFC. The end result of the parsing phase is now a `Grammar` which is a well structured rrule input (but not yet validated) and can further be used to construct a `RRuleSet`. There is still room for improvement to the parser, but these changes makes it better.

Another consequence of this PR is that `RRule` no longer implements `FromStr`. The reason for this is that in order to convert the `UNTIL` string to a `DateTime` we need to know the value of `DTSTART`, because these two values needs to be synced on timezones.

Fixes:
- #32 
- #58

### CODE REVIEW

**Comment by fmeringdal** [comment]
Hey @omid , would love to hear your thought about this rewrite of the parser if you have time :) No worries if you don't have time

File: rrule/src/parser/content_line/content_line_parts.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
+use crate::parser::{regex::get_property_name, ParseError};
+
+use super::PropertyName;
+
+#[derive(Debug, PartialEq, Clone)]
+pub(crate) struct ContentLineCaptures {
+    pub property_name: PropertyName,
+    pub parameters: Option<String>,
```
Comment (omid) [inline rrule/src/parser/content_line/content_line_parts.rs:8]: I expect `parameters` be a hashmap (or any other kind of iter). Cannot we parse parameters at this stage, in `get_content_line_parts`?

File: rrule/src/parser/content_line/content_line_parts.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    pub value: String,
```
Comment (omid) [inline rrule/src/parser/content_line/content_line_parts.rs:9]: I think all strings here in this struct can be `&str` to avoid copying. But it's not a biggie.

File: rrule/src/parser/content_line/parameters.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
+use std::{collections::HashMap, hash::Hash, str::FromStr};
+
+use crate::parser::ParseError;
+
+/// Parses a string of semi colon seperated key value pairs into a HashMap with
```
Comment (omid) [inline rrule/src/parser/content_line/parameters.rs:5]: typo: `semicolon`

File: rrule/src/parser/datetime.rs [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+    // Apply timezone from `TZID=` part (if any), else set datetime as UTC
+    let datetime_with_timezone = datetime.with_timezone(&tz.unwrap_or(UTC));
+
+    Ok(datetime_with_timezone)
+}
+
+/// Attempts to convert a `str` to a `Weekday`.
+pub(crate) fn str_to_weekday(d: &str) -> Result<Weekday, ParseError> {
+    let day = match &d.to_uppercase()[..] {
+        "MO" => Weekday::Mon,
```
Comment (omid) [inline rrule/src/parser/datetime.rs:121]: Not your changes, but I think it's better to define ALLL static ASCII strings as binary `b"MO"`. And instead of `&str`, pass `&[u8]` around.
Sure it should be another PR, it's a big and important change!

File: rrule/src/parser/regex.rs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+/// is optional. It returns [`ParsedDateString`].
+pub(crate) fn parse_datestring(val: &str) -> Result<ParsedDateString, ()> {
+    let captures = DATESTR_RE.captures(val).ok_or(())?;
+
+    let year = captures
+        .get(1)
+        .ok_or(())?
+        .as_str()
+        .parse()
+        .map_err(|_| ())?;
```
Comment (omid) [inline rrule/src/parser/regex.rs:45]: I don't like returning an empty tuple as error, maybe a new error type, but not an empty tuple.

If you don't want to repeat the error type, you can have a code like this:
```
captures
        .get(1)
        .and_then(|v| v.as_str().parse().ok())
        .ok_or(())?;
```

File: rrule/src/core/rrule.rs [hunk 6]
```diff
@@ -210,9 +175,7 @@ fn weekday_to_str(d: Weekday) -> String {
 /// It has two stages, based on the attached type, `Validated` or `Unvalidated`.
 /// - `Unvalidated`, which is the raw string representation of the RRULE
 /// - `Validated`, which is when the RRule has been parsed and validated, based on the start date
-#[cfg_attr(feature = "serde", serde_as)]
 #[derive(Debug, Clone, PartialEq, Eq)]
-#[cfg_attr(feature = "serde", derive(DeserializeFromStr, SerializeDisplay))]
 pub struct RRule<Stage = Validated> {
```
[the remark this answers is not in the document]
Comment (fmeringdal) [reply rrule/src/core/rrule.rs:179]: It is the same reason as why why have to remove `FromStr` implementation for `RRule` as explained in the description. Is there a reason why someone would go from a text input to an `RRule` rather than an `RRuleSet`?

File: rrule/src/parser/datetime.rs (hunk 4 above)
Comment (fmeringdal) [reply rrule/src/parser/datetime.rs:121]: Hmm, that is interesting and definitely something we can look into in another PR!

File: rrule/src/core/rrule.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/rrule.rs:179]: Oh, I missed the description 🙈 

Cannot we change the `UNTIL` timezone, after validation? Or return error if timezone of `UNTIL` is different from timezone of `DTSTART`?

PS: I exactly use `RRule::from_str()` in our application :D

File: rrule/src/parser/content_line/content_line_parts.rs (hunk 1 above)
Comment (fmeringdal) [reply rrule/src/parser/content_line/content_line_parts.rs:8]: I forgot that `value` is not actually a `HashMap`, for example `DTSTART;VALUE=DATE:19970101` where `value` is  `19970101` which is just a string.

I attempted to make the `parameters` a `HashMap<K, &str>`, where K is `DateParameter` (for `RDATE`,`DTSTART` etc ) or `RuleParameter` (for `RRULE`, `EXRULE`). The problem is that the generic K has to be provided by the caller of `get_content_line_parts`, which doesn't know what kind of line it is yet so can't possibly know what K should be!

Another solution would be to use `HashMap<&str, &str>` and then later convert that to `HashMap<DateParameter, &str>`, but I think at that point it is just easier to keep `parameters` as a `&str`.

Comment (omid) [reply rrule/src/parser/content_line/content_line_parts.rs:8]: Since the only caller of this is `Grammer`. I think it will be fine if we use `&str` everywhere. At least for this iteration.

File: rrule/src/parser/content_line/content_line_parts.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+/// Get the property name, parameters and values of a content line.
+pub(crate) fn get_content_line_parts(val: &str) -> Result<ContentLineCaptures, ParseError> {
```
Comment (omid) [inline rrule/src/parser/content_line/content_line_parts.rs:13]: Cannot it be a `impl FromStr`?

File: rrule/src/parser/regex.rs [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+#[derive(Debug, PartialEq)]
+pub(crate) struct ParsedDateStringTime {
+    pub hour: u32,
+    pub min: u32,
+    pub sec: u32,
+}
+
+/// Parses a date string with format `YYYYMMDD(THHMMSSZ)` where the part in paranthesis
+/// is optional. It returns [`ParsedDateString`].
+pub(crate) fn parse_datestring(val: &str) -> Result<ParsedDateString, ()> {
```
Comment (omid) [inline rrule/src/parser/regex.rs:37]: Cannot it be a `impl FromStr`?

File: rrule/src/core/rrule.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (fmeringdal) [reply rrule/src/core/rrule.rs:179]: Hmm, interesting. Can you provide some more details on how you use it and why you can't use `RRuleSet`? Is it a user input where you don't know the start date?

It seems like most other rrule implementations don't really follow the `UNTIL` / `DTSTART` sync according to the RFC, so I am kind of open to ignoring it if it proves to be a big pain to implement.

File: rrule/src/parser/content_line/content_line_parts.rs (hunk 7 above)
Comment (fmeringdal) [reply rrule/src/parser/content_line/content_line_parts.rs:13]: Yes, you are right, it _can_ be. I am a little bit inconsistent in my coding when choosing between `impl FromStr for Type` and `fn method(&str) -> Type`. Sometimes I want to add some docs to the implementation and give the method a specific name and therefore decide to go for `fn method..`, other times it feels more natural to implement `FromStr`. 

Do you follow any specific rules when choosing between the two approaches?

File: rrule/src/core/rrule.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/rrule.rs:179]: I cannot provide more details, but yes, something like that, the input is like just the `RRule` line, without the `DTSTART`.

File: rrule/src/parser/content_line/content_line_parts.rs (hunk 7 above)
Comment (omid) [reply rrule/src/parser/content_line/content_line_parts.rs:13]: If there is something in the standard library, I use that, because it's well known to developers and the behavior is exactly as they expected. That's it for me.

Comment (fmeringdal) [reply rrule/src/parser/content_line/content_line_parts.rs]: Having changed `ContentLineCaptures` to contain `&str` references it is not possible to implement `FromStr`, because the `FromStr` trait does not allow tying the input lifetime to the output lifetime. Anonymous lifetime doesn't seem to work either. Ref: https://stackoverflow.com/questions/28931515/how-do-i-implement-fromstr-with-a-concrete-lifetime

File: rrule/src/parser/regex.rs (hunk 8 above)
Comment (fmeringdal) [reply rrule/src/parser/regex.rs:37]: In this specific case, I would like to keep it as a function. There are many possible string formats for a date-time, so implementing `FromStr` makes it unclear exactly what format it expects, so I would like to keep it more explicit with a function with docs.

File: rrule/src/parser/content_line/content_line_parts.rs (hunk 7 above)
Comment (omid) [reply rrule/src/parser/content_line/content_line_parts.rs]: Yes, you are right. Then I suggest an `impl ContentLineCaptures` with a method like `from_str` or `from_blahblah_str`. Same for `ParsedDateString`.

Comment (fmeringdal) [reply rrule/src/parser/content_line/content_line_parts.rs]: Completely agree 👍

File: rrule/src/parser/content_line/content_line_parts.rs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                let mut parameters = None;
+                if line.starts_with(&format!("{};", property_name)) {
+                    let only_colon_idx = line.find(':');
+                    if let Some(only_colon_idx) = only_colon_idx {
+                        parameters =
+                            Some(&line[property_name.to_string().len() + 1..only_colon_idx]);
+                    }
+                }
+
+                Ok(ContentLineCaptures {
```
Comment (omid) [inline rrule/src/parser/content_line/content_line_parts.rs:34]: Will I be so picky if I say let's use `Self` here? 🙈

File: rrule/src/core/rrule.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (fmeringdal) [reply rrule/src/core/rrule.rs:179]: <details>
<summary>Here is how the RFC defines `UNTIL` and how it should sync with `DTSTART`</summary>
      The UNTIL rule part defines a DATE or DATE-TIME value that bounds
      the recurrence rule in an inclusive manner.  If the value
      specified by UNTIL is synchronized with the specified recurrence,
      this DATE or DATE-TIME becomes the last instance of the
      recurrence.  The value of the UNTIL rule part MUST have the same
      value type as the "DTSTART" property.  Furthermore, if the
      "DTSTART" property is specified as a date with local time, then
      the UNTIL rule part MUST also be specified as a date with local
      time.  If the "DTSTART" property is specified as a date with UTC
      time or a date with local time and time zone reference, then the
      UNTIL rule part MUST be specified as a date with UTC time.  In the
      case of the "STANDARD" and "DAYLIGHT" sub-components the UNTIL
      rule part MUST always be specified as a date with UTC time.  If
      specified as a DATE-TIME value, then it MUST be specified in a UTC
      time format.  If not present, and the COUNT rule part is also not
      present, the "RRULE" is considered to repeat forever.
</details>

The key thing is that the `VALUE` (i.e. either `DATE` or `DATETIME`) parameter needs to be synced and that if `DTSTART` is a `DATE` with local time then `UNTIL` must also be in local time. The validator phase only sees the `DateTime` type for `until` and `dtstart` and cannot based on that type validate those requirements (`DateTime` does not tell us if the original text representation was in local time nor if it was a `DATE` originally). It should be possible to reorganize the code such that the we can pass this information to the validator, but I feel that it would be non-trivial and also add tech debt.

I understand that you can't share details about your company's code, but I can't really understand why the `impl FromStr for RRule` is useful. Could you either add a dummy start date to the text input or store the input as a string until you have access to the start date?

[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/rrule.rs:179]: OK, it's fine.
Most probably you are right, it's better to touch some bits in the app.
